modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/BinaryTupleBuilder.java

```diff
@@ -315,8 +315,24 @@ public BinaryTupleBuilder appendNumber(BigInteger value) {
      * @return {@code this} for chaining.
      */
     public BinaryTupleBuilder appendDecimalNotNull(BigDecimal value, int scale) {
-        // TODO IGNITE-21745: Inefficient serialization, many bytes wasted to store small values when scale is big.
-        putBytes(value.setScale(scale, RoundingMode.HALF_UP).unscaledValue().toByteArray());
+        if (value.scale() > scale) {
+            value = value.setScale(scale, RoundingMode.HALF_UP);
+        }
+
+        value = value.stripTrailingZeros();
+
+        // See CatalogUtils.MAX_DECIMAL_SCALE = Short.MAX_VALUE
+        if (value.scale() > Short.MAX_VALUE) {
+            throw new BinaryTupleFormatException("Decimal scale is too large: " + value.scale() + " > " + Short.MAX_VALUE);
+        }
+
+        if (value.scale() < Short.MIN_VALUE) {
+            throw new BinaryTupleFormatException("Decimal scale is too small: " + value.scale() + " < " + Short.MIN_VALUE);
+        }
```

Comment:
Can there be a situation when `scale()` fits the range before `stripTrailingZeros()` and doesn't after it?

Reply:
Good catch, you are right. As the javadoc says, stripping the trailing zeros from `600.0 [value=6000, scale=1]` yields `6E2  [value=6, scale=-2]`.

We don't want to remove `stripTrailingZeros` completely - it produces a more compact representation. So I've added a scale check.

Reply:
Added a test too (`BinaryTupleTest.maxDecimalScaleTest`) - it failed before the fix and now works.